tests/kafkatest/tests/streams/streams_relational_smoke_test.py

```diff
@@ -0,0 +1,123 @@
+# limitations under the License.
+
+
+from ducktape.mark import matrix
+from ducktape.mark.resource import cluster
+from ducktape.utils.util import wait_until
+from kafkatest.services.streams import StreamsTestBaseService
+from kafkatest.tests.kafka_test import KafkaTest
+
+
+class StreamsRelationalSmokeTestService(StreamsTestBaseService):
+    def __init__(self, test_context, kafka, mode, nodeId):
+        super(StreamsRelationalSmokeTestService, self).__init__(
+            test_context,
+            kafka,
+            "ignore",
+            "ignore"
+        )
+        self.mode = mode
+        self.nodeId = nodeId
+        self.log4j_template = 'log4j_template.properties'
+
+    def start_cmd(self, node):
+        return "( export KAFKA_LOG4J_OPTS=\"-Dlog4j.configuration=file:%(log4j)s\"; " \
+               "INCLUDE_TEST_JARS=true %(kafka_run_class)s org.apache.kafka.streams.tests.RelationalSmokeTest " \
+               " %(mode)s %(kafka)s %(nodeId)s %(state_dir)s" \
+               " & echo $! >&3 ) 1>> %(stdout)s 2>> %(stderr)s 3> %(pidfile)s" % {
+                   "log4j": self.LOG4J_CONFIG_FILE,
+                   "kafka_run_class": self.path.script("kafka-run-class.sh", node),
+                   "mode": self.mode,
+                   "nodeId": self.nodeId,
+                   "kafka": self.kafka.bootstrap_servers(),
+                   "state_dir": self.PERSISTENT_ROOT,
+                   "stdout": self.STDOUT_FILE,
+                   "stderr": self.STDERR_FILE,
+                   "pidfile": self.PID_FILE
+               }
+
+    def start_node(self, node):
+        node.account.mkdirs(self.PERSISTENT_ROOT)
+        node.account.create_file(self.LOG4J_CONFIG_FILE,
+                                 self.render("log4j_template.properties", log_file=self.LOG_FILE))
+
+        self.logger.info("Starting process on " + str(node.account))
+        node.account.ssh(self.start_cmd(node))
+
+        if len(self.pids(node)) == 0:
+            raise RuntimeError("No process ids recorded")
+
+    def await_command(self, command):
+        wait_until(lambda: self.node.account.ssh(command, allow_fail=True),
+                   timeout_sec=60,
+                   err_msg="Command [%s] never passed in the timeout"
+                   )
+
+
+class StreamsRelationalSmokeTest(KafkaTest):
+    """
+    Simple test of Kafka Streams.
+    """
+
+    def __init__(self, test_context):
+        super(StreamsRelationalSmokeTest, self).__init__(test_context, num_zk=1, num_brokers=3, topics={
+            'in-article': {'partitions': 3, 'replication-factor': 1},
+            'in-comment': {'partitions': 5, 'replication-factor': 1},
+            'out-augmented-article': {'partitions': 3, 'replication-factor': 1},
+            'out-augmented-comment': {'partitions': 5, 'replication-factor': 1}
+        })
+        self.test_context = test_context
+
+    @cluster(num_nodes=8)
+    @matrix(crash=[False, True])
+    def test_streams(self, crash):
```

Comment:
One thing to note is the avoidance of log_monitor, which doesn't work the way that it seems like we thought it works. Instead, we're structuring our ssh/grep commands to make the minimum ordering assumptions and be correct by looking for unique events.

Reply:
Could you elaborate a bit more about `log_monitor` why it does not work as expected?

Reply:
The log monitor looks up the current rail position of the file just once when you create it. Subsequently, each grep invocation uses the exact same starting point in the file. 

I’ve noticed that some of our tests seem to assume that the monitor maintains a cursor in the file, but it does not. 

The one nice thing about the monitor is that within it, you’re guaranteed not to see any log messages from before the time you created it, but in practice there are still too many opportunities to miss something or make an incorrect assumption in the test. 

I think a more robust approach is to just expect to search the entire log each time and figure out how to phrase the query as an invariant.